File: 3664_two_letter_card_game/Solution.scala
// LeetCode 3664 - Two-Letter Card Game
// https://leetcode.com/problems/two-letter-card-game/

object Solution {
  private def pairGroup(arr: Array[Int]): Array[Int] = {
    var total = 0
    var mx = 0
    var i = 0
    while (i < 26) {
      total += arr(i)
      mx = math.max(mx, arr(i))
      i += 1
    }
    var pairs = total / 2
    if (total - mx < pairs) pairs = total - mx
    Array(pairs, total - 2 * pairs)
  }

  def score(cards: Array[String], x: Char): Int = {
    var xx = 0
    val left = new Array[Int](26)
    val right = new Array[Int](26)
    for (c <- cards) {
      val a = c.charAt(0)
      val b = c.charAt(1)
      if (a == x && b == x) xx += 1
      else if (a == x) left(b - 'a') += 1
      else if (b == x) right(a - 'a') += 1
    }
    val lp = pairGroup(left)
    val rp = pairGroup(right)
    var ans = lp(0) + rp(0)
    val rem = lp(1) + rp(1)
    val use = math.min(xx, rem)
    ans += use
    xx -= use
    ans += xx / 2
    ans
  }
}

File: 3664_two_letter_card_game/solution.rs
// LeetCode 3664 - Two-Letter Card Game
// https://leetcode.com/problems/two-letter-card-game/

impl Solution {
    pub fn score(cards: Vec<String>, x: char) -> i32 {
        let mut xx = 0;
        let mut left = [0i32; 26];
        let mut right = [0i32; 26];
        for c in cards {
            let b = c.as_bytes();
            let a = b[0] as char;
            let d = b[1] as char;
            if a == x && d == x {
                xx += 1;
            } else if a == x {
                left[(d as u8 - b'a') as usize] += 1;
            } else if d == x {
                right[(a as u8 - b'a') as usize] += 1;
            }
        }
        let pair_group = |arr: &[i32; 26]| -> (i32, i32) {
            let mut total = 0;
            let mut mx = 0;
            for &v in arr {
                total += v;
                mx = mx.max(v);
            }
            let mut pairs = total / 2;
            if total - mx < pairs {
                pairs = total - mx;
            }
            (pairs, total - 2 * pairs)
        };
        let (lp, lr) = pair_group(&left);
        let (rp, rr) = pair_group(&right);
        let mut ans = lp + rp;
        let rem = lr + rr;
        let use_xx = xx.min(rem);
        ans += use_xx;
        xx -= use_xx;
        ans += xx / 2;
        ans
    }
}
